### DISCUSSION TITLE
Terminal Multiplexer idea + prototype

### DISCUSSION THREAD

**Discussion by ceccopierangiolieugenio** [ceccopierangiolieugenio/pyTermTk#207] [Ideas]
> Hello, I would like to ask if there are plans for making a TWM with this, for example, I like [twin](https://github.com/cosmos72/twin) but it comes with a lot of problems due to being written in ncurses and C. Problems such as, mouse and keyboard handling and being hard to change source code due to hard to understand code.
> 
> I looked around pyTermTk's code and absolutely loved it. It is the easiest to understand Python TUI library in my opinion but I can't wrap my head around writing a WM around it. Perhaps I need some more time with it. First I thought it would be cool to write a code editor in it but I realized that already exists and at a good state. Now I don't have many projects in my mind that I can do with this, that isn't as extreme as writing a window manager.
> 
> I also want to make some suggestions, if you were to create this, I made a mockup: ![idea](https://private-user-images.githubusercontent.com/36235265/305665717-eb4ed193-7f1d-454b-81bd-f297a39a9f76.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1OTA3MjgsIm5iZiI6MTcwODU5MDQyOCwicGF0aCI6Ii8zNjIzNTI2NS8zMDU2NjU3MTctZWI0ZWQxOTMtN2YxZC00NTRiLTgxYmQtZjI5N2EzOWE5Zjc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIyVDA4MjcwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU3MmE1NjViOWNiZjI4OTZjZGFmODFlMmNjNDYzNGY4YmNhZWUwMDY3NzVkMjE0ODhiOTAyMTBjNTFiNjBjYmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.IeJscRHh2S_iL9EzBlp2TxZ2WIuR93JClGbXt_YP4N8)
> 
> Also, I'm not a great programmer but if you want to discuss this ideavise, I'm always open for it!
> 
> Thank you for all the great work so far!

**Comment by agoodfellow123** [comment]
Hello, I would like to ask if there are plans for making a TWM with this, for example, I like [twin](https://github.com/cosmos72/twin) but it comes with a lot of problems due to being written in ncurses and C.
Problems such as, mouse and keyboard handling and being hard to change source code due to hard to understand code.

I looked around pyTermTk's code and absolutely loved it. It is the easiest to understand Python TUI library in my opinion but I can't wrap my head around writing a WM around it. Perhaps I need some more time with it. First I thought it would be cool to write a code editor in it but I realized that already exists and at a good state. Now I don't have many projects in my mind that I can do with this, that isn't as extreme as writing a window manager.

I also want to make some suggestions, if you were to create this, I made a mockup:
![idea](https://github.com/ceccopierangiolieugenio/pyTermTk/assets/36235265/eb4ed193-7f1d-454b-81bd-f297a39a9f76)

Also, I'm not a great programmer but if you want to discuss this ideavise, I'm always open for it!

Thank you for all the great work so far!

**Reply by ceccopierangiolieugenio** [reply]
your idea should be straightforward to implement.

I already made a WM for fun in:
https://www.youtube.com/watch?v=p1MiZOLCFOY
you can try with:
```bash
git clone https://github.com/ceccopierangiolieugenio/pyTermTk
pyTermTk/multiplexers/workbench/main.py
```

About the editor I started long ago working on 
https://github.com/ceccopierangiolieugenio/ttkode
a code editor with a vscode look and feel.
Right now I am working to a new layout manager required for for my next updates on the editor.

**Reply by agoodfellow123** [reply]
Hmm yes, I also tried the basic multiplexer, perhaps I can make it on that code base. Thank you.

**Reply by ceccopierangiolieugenio** [reply]
Right now I am busy with the logic of the appTemplate.
In the next days I can try to prototype your idea.
So far I think everything can be implemented except the clock at the center (no problem having it on the right/left)

**Reply by agoodfellow123** [reply]
That would be great! Thank you. I'll try to customize what you made. Maybe even improve, for example adding drag and drop tiling to sides and the top. Please keep me updated.

**Reply by ceccopierangiolieugenio** [reply]
I made a quick demo:

https://github.com/ceccopierangiolieugenio/pyTermTk/assets/8876552/ed950f88-59ec-4dd3-84c7-790fc5baf205

Here are the sources, feel free to ask me anything, there are still may cases to be implemented, i.e. the interaction between terminal focus and list, or the terminal close routine.
The maximize/minimize may not works properly with the current pyTermTk, I noticed a bug and I will provide a fix in the latest git commit.

```python
#!/usr/bin/env python3

import TermTk as ttk

class TerminalList(ttk.TTkVBoxLayout):
    def __init__(self, *args, **kwargs):
        self.newTerminal = ttk.pyTTkSignal()
        self._listTerminals = []
        super().__init__(*args, **kwargs)
        self._newBtn = ttk.TTkButton(text="New Term...", maxHeight=3, border=True)
        self._tlist   = ttk.TTkList(dragDropMode=ttk.TTkK.AllowDragDrop)

        self.addWidgets([self._newBtn, self._tlist])
        self._newBtn.clicked.connect(self.newTerminal.emit)

        self._tlist.itemClicked.connect(self._itemClicked)

    def addTerminal(self, title, win, term):
        data = {'title':title, 'win':win, 'term':term}
        self._tlist.addItem(title, data)
        self._tlist.setCurrentRow(len(self._listTerminals))
        self._listTerminals.append(data)

    @ttk.pyTTkSlot(ttk.TTkAbstractListItem)
    def _itemClicked(self, item):
        term = item.data()['term']
        term.setFocus()
        term.raiseWidget()

class VirtualDesktop(ttk.TTkSplitter):
    # This class include a simple splitter layout
    # with a terminal lists in the left side
    # and an empty layout on the right side where
    # where the terminal windows are placed
    def __init__(self, *args, **kwargs):
        # Used to define a unique name
        self._enumerate = 0
        super().__init__(*args, **kwargs)
        self._terminalList = TerminalList()
        self._mainLayout   = ttk.TTkLayout()
        self.addItem(self._terminalList, size=30)
        self.addItem(self._mainLayout)
        self._terminalList.newTerminal.connect(self._newTerminal)

    @ttk.pyTTkSlot()
    def _newTerminal(self):
        #
        self._enumerate += 1
        name =  f"Terminal {self._enumerate}"

        # Find a new position
        w,h = self.size()
        newX, newY = 0,0
        positions = [w.pos() for w in self._mainLayout.children()]
        while (newX,newY) in positions: newX+=2 ; newY+=3
        # Avoid the new terminal to be placed outside the frame
        newX = min(newX, w-10)
        newY = min(newY, h-5)

        win = ttk.TTkWindow(title=name,
                            pos=(newX,newY), size=(80+2,24+5),
                            flags = ttk.TTkK.WindowFlag.WindowMinMaxButtonsHint | ttk.TTkK.WindowFlag.WindowReduceButtonHint ,
                            layout=ttk.TTkGridLayout())
        term = ttk.TTkTerminal(parent=win)
        self._terminalList.addTerminal(name, win, term)

        th = ttk.TTkTerminalHelper(term=term)
        th.terminalClosed.connect(win.close)

        term.bell.connect(lambda : ttk.TTkLog.debug("BELL!!! 🔔🔔🔔"))
        term.titleChanged.connect(win.setTitle)

        # term.textSelected.connect(clipboard.setText)
        # term.textSelected.connect(_textSelected)
        win.closed.connect(term.close)
        self._mainLayout.addWidget(win)
        term.setFocus()
        term.raiseWidget()

        th.runShell()

class MainDesktop(ttk.TTkGridLayout):
    def __init__(self, *args, **kwargs):
        # Used to define a unique name
        self._enumerate = 0
        super().__init__(*args, **kwargs)
        self._tabVirtDesk = ttk.TTkTabWidget(border=False)
        self.addWidget(self._tabVirtDesk)
        self._tabVirtDesk.addMenu("New...").menuButtonClicked.connect(self._addVirtualDesktop)
        self._addVirtualDesktop()

    @ttk.pyTTkSlot()
    def _addVirtualDesktop(self):
        self._enumerate += 1
        self._tabVirtDesk.addTab(VirtualDesktop(),f"  {self._enumerate}  ")


root = ttk.TTk(layout=MainDesktop())
root.mainloop()
```

**Reply by agoodfellow123** [reply]
Sorry for the late response! It looks absolutely great! I'll tinker around a bit. Thank you!

**Reply by agoodfellow123** [reply]
Hmm, i hit a wall, `def size(self):     return self._w, self._h` seems to take two agruments but when I enter `self.size(ttk.TTkLayoutItem(4,3))` under `class TerminalList(ttk.TTkVBoxLayout):` gives me `TypeError: TTkLayoutItem.size() takes 1 positional argument but 2 were given` error. Also I changed the line 90 made it like this to ensure user starts with 10 windows. 
```
        for i in range(10):
            self._addVirtualDesktop()
```
According to a lot of people on unixpron 10 is the perfect number. if user doesn't like it they can change the code or perhaps there will be a config, it's all up to you chef!

I must apologize for my lack of understanding. I learned coding at an early age but never been able to grasp oop.

**Reply by ceccopierangiolieugenio** [reply]
about spawning 10 desktops, there is nothing wrong with it.
did you have issues with it?

about the size.
this is an object method.
you don't have to pass any arguments and it return a tuple with the dimensions of the widget/layout:

i.e.:
```python
exampleLayoutItem = ttk.TTkLayoutItem()
w,h = vexampleLayoutItem.size()
```
or

https://github.com/ceccopierangiolieugenio/pyTermTk/blob/d122b8fd858e363f71c5a5dcadf283ba2e719baa/TermTk/TTkWidgets/apptemplate.py#L417-L418

or

https://github.com/ceccopierangiolieugenio/pyTermTk/blob/d122b8fd858e363f71c5a5dcadf283ba2e719baa/TermTk/TTkWidgets/menu.py#L247-L253

**Reply by agoodfellow123** [reply]
I had no issues with spawning 10 desktops, I just thought if we removed the "new desktop" button it would look cleaner.

Thank you for the code, I will look into it. Also If I remember correctly keyboard shortcuts still doesn't work on pytermtk. So again for less busy looking interface I'd say adding the "new terminal" button to the very bottom of that panel might be a better look. And instead of new terminal text we can do just a "+" sign like this:

![image](https://github.com/ceccopierangiolieugenio/pyTermTk/assets/36235265/966086ec-8bdb-4043-9a6a-dd3353135dd1)

**Reply by ceccopierangiolieugenio** [reply]
shortcuts are supported through the TTkShortcut helper, i.e.

https://github.com/ceccopierangiolieugenio/pyTermTk/blob/6e225d72e1ba1918aa4d6c9b5844fa4ecad0a192/ttkDesigner/app/designer.py#L188-L189

I introduced it recently and it is still in early development and I noticed some corner cases where it doesn't work as expected, i need to fix the key event handlers of some widgets because the list widget and textedit for example capture the input and don't allow the shortcuts to properly works

**Reply by ceccopierangiolieugenio** [reply]
hi @agoodfellow123,
I'd like to include your idea of terminal multiplexer in the multiplexer folder so I can split it in multiple files and make it more feature rich, and configurable.
About the side panel with the list of terminals, at the moment , my previous code has a different task/terminals panel per desktop.
Instead, were you thinking of having a side panel with all the terminal collected in between desktops, so you can jump straight to the desktop/terminal on a single click?
I think we can make it a tree widget divided per desktops, instead of a simple list to make it easier to identify the terminal required.

**Reply by agoodfellow123** [reply]
Ah, no. I didn't mean shared desktop terminals. Each desktop should have their own terminals in my opinion.

Tree view is ok but kills serious screen real estate, that's why I recommended icons only terminal task bar. Perhaps in the future it would be possible to add tooltips.

Also, I'm honored by your decision to add this idea to the multiplexers!
